PixelHunter1995/WalkingAreaLib/WalkingArea.cs

```diff
@@ -31,28 +33,113 @@ public void Draw(GraphicsDeviceManager graphics, SpriteBatch spriteBatch, double
             partition.Draw(graphics, spriteBatch, scaling, sceneWidth);
         }
 
-        public Vector2 GetNextPosition(Vector2 clickPosition, Vector2 currentPosition)
+        private Vector2 ClosestPositionInWalkingArea(Vector2 inPosition)
         {
-            // clickPosition outside of PolygonPartition
-            if (!partition.Contains(clickPosition))
+            if (partition.Contains(inPosition))
             {
-                // TODO: Get closest position on edge not just closest vertex
-                Vector2 closestVertex = Vector2.Zero;
-                double closestDistance = double.MaxValue;
-                foreach (var polygon in partition.Polygons)
+                return inPosition;
+            }
+
+            Vector2 closestPoint = Vector2.Zero;
+            double closestDistance = double.MaxValue;
+            foreach (var polygon in partition.Polygons)
+            {
+                (Vector2 closest, double distance) = polygon.ClosestPositionInPolygon(inPosition);
+
+                if (distance < closestDistance)
                 {
-                    (Vector2 closest, double distance) = polygon.GetClosestVertex(clickPosition);
+                    closestPoint = closest;
+                    closestDistance = distance;
+                }
+            }
+
+            return closestPoint;
+        }
+
+        public Vector2 GetNextPosition(Vector2 clickPosition, Vector2 currentPosition)
+        {
+            clickPosition = ClosestPositionInWalkingArea(clickPosition);
+            currentPosition = ClosestPositionInWalkingArea(currentPosition);
+
+            int clickIndex = partition.ContainingPolygonIndex(clickPosition);
+            int currentIndex = partition.ContainingPolygonIndex(currentPosition);
+
+            // Case 1: There's a straight line within the walking area
+            //         between the current position and the clicked position.
+            if (partition.ContainsLine(currentPosition, clickPosition))
+            {
+                return clickPosition;
+            }
+
+            // Case 2: Find adjacent polygon in correct direction.
+            int[] path = RunDijkstra(partition.Polygons.Count, partition.DistanceMatrix, currentIndex);
```

Comment:
Seems like we still do a lot of re-calculation, don't we? Though I suppose in most cases the condition above will have us return early.

On that note we could return even earlier if clickPosition and and currentPosition are equal.